src/java/com/puppetlabs/http/client/impl/StreamingAsyncResponseConsumer.java

```diff
@@ -0,0 +1,88 @@
+package com.puppetlabs.http.client.impl;
+
+import org.apache.http.Header;
+import org.apache.http.HeaderElement;
+import org.apache.http.HttpEntity;
+import org.apache.http.HttpResponse;
+import org.apache.http.client.entity.DeflateDecompressingEntity;
+import org.apache.http.client.entity.GzipDecompressingEntity;
+import org.apache.http.entity.BasicHttpEntity;
+import org.apache.http.nio.IOControl;
+import org.apache.http.nio.client.methods.AsyncByteConsumer;
+import org.apache.http.protocol.HttpContext;
+
+import java.io.IOException;
+import java.io.PipedInputStream;
+import java.io.PipedOutputStream;
+import java.nio.ByteBuffer;
+
+public class StreamingAsyncResponseConsumer extends AsyncByteConsumer<HttpResponse> {
+
+    private volatile HttpResponse response;
+    private volatile PipedOutputStream pos;
+    private volatile Deliverable<HttpResponse> promise;
+    private volatile Promise<IOException> ioExceptionPromise = new Promise<>();
+
+    public void setFinalResult(IOException ioException) {
+        ioExceptionPromise.deliver(ioException);
+    }
+
+    public StreamingAsyncResponseConsumer(Deliverable<HttpResponse> promise) {
+        this.promise = promise;
+    }
+
+    @Override
+    protected void onResponseReceived(final HttpResponse response) throws IOException {
+        PipedInputStream pis = new ExceptionInsertingPipedInputStream(ioExceptionPromise);
+        pos = new PipedOutputStream();
+        pos.connect(pis);
+        HttpEntity modifiedEntity = new BasicHttpEntity();
+        ((BasicHttpEntity) modifiedEntity).setContent(pis);
+        HttpEntity entity = response.getEntity();
```

Comment:
Seems to work well. As far as I can see we can use this code to handle all decompression rather than reserving it just for :unbuffered-stream. I've implemented it that way here (in a separate commit). If that seems too risky we can enable it only in the new code path.